File: Cpp/Example4-4-CalDistancePointAndLine/Example4-4-CalDistancePointAndLine.cpp
// Example4CalDistancePointAndLine.cpp : 此文件包含 "main"
// 函数。程序执行将在此处开始并结束。
//

#include <Eigen/Eigen>
#include <iostream>

using namespace Eigen;
using namespace std;

double CalDistancePointAndLine(Vector3d& point, Vector3d& lineBegin,
                               Vector3d& lineEnd) {
  //直线方向向量
  Vector3d n = lineEnd - lineBegin;

  //直线上某一点的向量到点的向量
  Vector3d m = point - lineBegin;

  return (n.cross(m)).norm() / n.norm();
}

double CalDistancePointAndLine1(Vector3d& point, Vector3d& lineBegin,
                               Vector3d& lineEnd) {
  double A = 1 / (lineEnd.x() - lineBegin.x());
  double B = -1 / (lineEnd.y() - lineBegin.y());
  double C = lineBegin.y() / (lineEnd.y() - lineBegin.y()) -
             lineBegin.x() / (lineEnd.x() - lineBegin.x());

  return abs(A * point.x() + B * point.y() + C) / sqrt(A * A + B * B);
}

int main() {
  Vector3d point(0.5, 0.6, 0);
  Vector3d O(1.0, 2.4, 0);
  Vector3d E(10.2, 11.5, 0);

  cout << "点到直线的距离为:" << CalDistancePointAndLine(point, O, E) << '\n';
  cout << "进行验算:" << CalDistancePointAndLine1(point, O, E) << '\n';
}

File: Python/example-4-4-cal-distance-point-and-line.py
import numpy as np

def cal_distance_point_and_line(point, lineBegin, lineEnd):
    # 直线方向向量
    n = lineEnd - lineBegin

    # 直线上某一点的向量到点的向量
    m = point - lineBegin

    return np.linalg.norm(np.cross(n, m)) / np.linalg.norm(n)


def cal_distance_point_and_line_1(point, lineBegin, lineEnd):
    A = 1 / (lineEnd[0] - lineBegin[0])
    B = -1 / (lineEnd[1] - lineBegin[1])
    C = lineBegin[1] / (lineEnd[1] - lineBegin[1]) - lineBegin[0] / (lineEnd[0] - lineBegin[0])

    return abs(A * point[0] + B * point[1] + C) / np.sqrt(A * A + B * B)


if __name__ == "__main__":
    point = np.array([0.5, 0.6, 0.0])
    O = np.array([1.0, 2.4, 0.0])
    E = np.array([10.2, 11.5, 0.0])

    print("点到直线的距离为:", cal_distance_point_and_line(point, O, E))
    print("进行验算:", cal_distance_point_and_line_1(point, O, E))